test/api/api_test.go

```diff
@@ -138,13 +140,77 @@ func TestWithDBSession(t *testing.T) {
 	}
 }
 
+// TestWithDBSession_RetryOnWriteConflict ensures that WithDBSession can
+// properly retry requests on MongoDB WriteConflict error.
+func TestWithDBSession_RetryOnWriteConflict(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	dbName := test.DBNameForTest(t.Name())
+	dep := lib.NewDependencyUserPutMongoDelay()
+	at, err := test.NewAccountsTester(dbName, dep)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if errClose := at.Close(); errClose != nil {
+			t.Error(errors.AddContext(errClose, "failed to close account tester"))
+		}
+	}()
+
+	// Ensure WithDBSession works with requests without bodies.
+	// This is a regression test. It panics with a nil pointer if we cannot
+	// properly handle requests with nil bodies.
+	testAPI, err := api.New(at.DB, nil, at.Logger, nil, nil)
+	if err != nil {
+		t.Fatal("Failed to instantiate API.", err)
+	}
+	handler := func(_ http.ResponseWriter, _ *http.Request, _ httprouter.Params) {}
+	req := (&http.Request{}).WithContext(context.Background())
+	testAPI.WithDBSession(handler)(nil, req, nil)
+
+	// Create a test user.
+	userEmail := types.NewEmail(t.Name() + "@siasky.net")
+	userPassword := t.Name() + "pass"
+	_, b, err := at.UserPOST(userEmail.String(), userPassword)
+	if err != nil {
+		t.Fatal(err, string(b))
+	}
+	defer func() {
+		_, _ = at.UserDELETE()
```

Comment:
so the account tester keeps track of the test user internally? I would've expected that you need to supply a username or id to the delete method.

Reply:
The tester keeps the user's auth token as internal state. This allows us to test authenticated and unauthenticated API calls. This is done for convenience. It can be replaced with something like `at.WithIdentity(token).UserDELETE()`, if needed.

What is the concern with the internal state?

Reply:
no concerns, will resolve